mllib/src/main/scala/org/apache/spark/ml/image/ImageSchema.scala

```diff
@@ -0,0 +1,229 @@
+        var offset = 0
+        for (h <- 0 until height) {
+          for (w <- 0 until width) {
+            val color = new Color(img.getRGB(w, h))
+
+            decoded(offset) = color.getBlue.toByte
+            decoded(offset + 1) = color.getGreen.toByte
+            decoded(offset + 2) = color.getRed.toByte
+            if (nChannels == 4) {
+              decoded(offset + 3) = color.getAlpha.toByte
+            }
+            offset += nChannels
+          }
+        }
+      }
+
+      // the internal "Row" is needed, because the image is a single dataframe column
+      Some(Row(Row(origin, height, width, nChannels, mode, decoded)))
+    }
+  }
+
+  /**
+   * :: Experimental ::
+   * Read the directory of images from the local or remote source
+   *
+   * @param path               Path to the image directory
+   * @param sparkSession       Spark Session
+   * @param recursive          Recursive path search flag
+   * @param numPartitions      Number of the dataframe partitions
+   * @param dropImageFailures  Drop the files that are not valid images from the result
+   * @param sampleRatio        Fraction of the files loaded
+   * @return                   Dataframe with a single column "image" of images;
+   *                           see ImageSchema for the details
+   */
+  @Since("2.3.0")
+  def readImages(path: String,
+                 sparkSession: SparkSession = null,
+                 recursive: Boolean = false,
+                 numPartitions: Int = 0,
+                 dropImageFailures: Boolean = false,
+                 sampleRatio: Double = 1.0): DataFrame = {
+    require(sampleRatio <= 1.0 && sampleRatio >= 0, "sampleRatio should be between 0 and 1")
+
+    val session = if (sparkSession != null) sparkSession else SparkSession.builder().getOrCreate
+    val partitions =
+      if (numPartitions > 0) {
+        numPartitions
+      } else {
+        session.sparkContext.defaultParallelism
+      }
+
+    val oldRecursiveFlag = RecursiveFlag.setRecursiveFlag(Some(recursive.toString), session)
+    val oldPathFilter: Option[Class[_]] =
+      if (sampleRatio < 1) {
+        SamplePathFilter.setPathFilter(Some(classOf[SamplePathFilter]), sampleRatio, session)
+      } else {
+        None
+      }
+
+    var result: DataFrame = null
+    try {
+      val streams = session.sparkContext.binaryFiles(path, partitions)
+        .repartition(partitions)
+
+      val images = if (dropImageFailures) {
+        streams.flatMap {
+          case (origin, stream) => decode(origin, stream.toArray)
+        }
+      } else {
+        streams.map {
+          case (origin, stream) => decode(origin, stream.toArray).getOrElse(invalidImageRow(origin))
+        }
+      }
+
+      result = session.createDataFrame(images, imageDFSchema)
+    }
+    finally {
+      // return Hadoop flags to the original values
+      RecursiveFlag.setRecursiveFlag(oldRecursiveFlag, session)
+      SamplePathFilter.unsetPathFilter(oldPathFilter, session)
```

Comment:
Seems this unconditionally changes default path filter if any when `sampleRatio` is less than 1. I guess we should avoid to unset it?

Reply:
Yes, when sample ratio is less than one we add a path filter to sample images.  We then reset back to the old path filter (if there was any) or remove the path filter, depending on whether one existed or not.  Sorry, I'm a bit confused though, why or how can we avoid to unset the path filter?  Or did you mean that we shouldn't be using a path filter?

Reply:
Oh, I meant `oldPathFilter` could be assigned `None` if `sampleRatio == 1`, then `unsetPathFilter` will unset path filter when the given old path filter is a `None`. May we unset any path filter instead of restoring it back in this case? So shall we skip calling `unsetPathFilter` if `sampleRatio == 1` because we didn't set it up?

Reply:
"Oh, I meant oldPathFilter could be assigned None if sampleRatio == 1, then unsetPathFilter will unset path filter when the given old path filter is a None."

yes, you are correct, oldPathFilter is set to None if sampleRatio == 1 based on the code:

    val oldPathFilter: Option[Class[_]] =
      if (sampleRatio < 1) {
        SamplePathFilter.setPathFilter(Some(classOf[SamplePathFilter]), sampleRatio, session)
      } else {
        None
      }

"May we unset any path filter instead of restoring it back in this case?"
You're correct, we should not be removing the path filter in this case.

"So shall we skip calling unsetPathFilter if sampleRatio == 1 because we didn't set it up?"
I see, yes, it does seem like we shouldn't be unsetting anything in this case, I've updated the code.  Good catch!